packages/pds/src/api/proxy.ts

```diff
@@ -0,0 +1,89 @@
+import * as express from 'express'
+import AtpAgent from '@atproto/api'
+import { Headers, XRPCError } from '@atproto/xrpc'
+import { InvalidRequestError, UpstreamFailureError } from '@atproto/xrpc-server'
+import AppContext from '../context'
+
+export const proxy = async <T>(
+  ctx: AppContext,
+  pdsDid: string | null | undefined,
+  fn: (agent: AtpAgent) => Promise<T>,
+): Promise<T | null> => {
+  if (isThisPds(ctx, pdsDid)) {
+    return null // skip proxying
+  }
+  const accountService = ctx.services.account(ctx.db)
+  const pds = pdsDid && (await accountService.getPds(pdsDid))
+  if (!pds) {
+    throw new UpstreamFailureError('unknown pds')
+  }
+  // @TODO reuse agents
```

Comment:
yeah unfortunately this really is necessary

Reply:
My plan is for this to go away when we incorporate #1730 😈